### DISCUSSION TITLE
How to tell if hand tracking is enabled?

### DISCUSSION THREAD

**Discussion by imjusthere** [microsoft/MixedRealityToolkit-Unity#11525] [Q&A]
I'm making a tutorial for the player and I need to change it for whether they are using hand tracking or not. What is the best way to determine if the device is using hand tracking or controllers?

**Comment by XRDevIEEE** [comment]
why not ask the user which they prefer

**Reply by imjusthere** [reply]
MRTK3 automatically adjusts to hand tracking if it doesn't find the controllers. I just need to know how to check to see which it is doing.

**Reply by AMollis** [reply]
Hi @imjusthere , are you wanting to know if the platform supports articulated hands? Or if a given interactor is backed by a hardware controller or articulated hands?

**Reply by imjusthere** [reply]
Neither. I'm developing for the Quest 2 and it is working fine. If it detects controller movement it uses controllers otherwise it's switching to hand tracking. I'm looking for the best way to determine which one it's using so I can change the tutorial content to match hand tracking or controllers.

**Reply by AMollis** [reply]
@imjusthere have you tried querying HandsAggregator for a hand joints?  `TryGetJoint` should fail if an articulated hand isn't being tracked.

```
        private bool IsHandTracked(Handedness hand)
        {
            // Early out if the hand isn't a valid XRNode
            // (i.e., Handedness.None or Handedness.Both)
            XRNode? node = hand.ToXRNode();
            if (!node.HasValue) { return false; }
            return XRSubsystemHelpers.HandsAggregator != null &&
                   XRSubsystemHelpers.HandsAggregator.TryGetJoint(TrackedHandJoint.Palm, node.Value, out HandJointPose pose);
        }
```

**Reply by AMollis** [reply]
@imjusthere  you might also be able to handle when a new input device is detected.  Then query the type of the device to determine if the system is using a hand or hardware controller

```
        protected void OnEnable()
        {
            UnityEngine.XR.InputDevices.deviceConnected += InputDevices_deviceConnected;
        }

        private void InputDevices_deviceConnected(UnityEngine.XR.InputDevice device)
        {
            // check type of device.
        }
```

Another option is to handle the `isTracked` input action property. You can also take a look at MRTK3's [ControllerVisualer](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/ef706f1f31cfc766a633ea2608456ad198484165/com.microsoft.mrtk.input/Visualizers/ControllerVisualizer/ControllerVisualizer.cs#LL55C1-L55C62) and follow that example.
